core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -1201,142 +1101,214 @@ object ZManaged {
+     * eagerly (with the [[Exit]] value with which the scope has ended) and
+     * no ReleaseKey will be returned.
      */
-    def remove: UIO[Set[Exit[Any, Any] => URIO[R, Any]]]
+    def addIfOpen(finalizer: Finalizer): UIO[Option[ReleaseKey]]
 
     /**
-     * Replaces the finalizers associated with this scope and returns them.
-     * If the finalizers associated with this scope have already been run this
-     * finalizer will be run immediately.
+     * Retrieves the finalizer associated with this key.
      */
-    def replace(finalizer: Exit[Any, Any] => URIO[R, Any]): URIO[R, Set[Exit[Any, Any] => URIO[R, Any]]]
+    def get(key: ReleaseKey): UIO[Option[Finalizer]]
 
     /**
      * Runs the specified finalizer and removes it from the finalizers
      * associated with this scope.
      */
-    def run(finalizer: Exit[Any, Any] => URIO[R, Any])(exit: Exit[Any, Any]): URIO[R, Unit]
+    def release(key: ReleaseKey, exit: Exit[Any, Any]): UIO[Any]
+
+    /**
+     * Runs the finalizers associated with this scope using the specified
+     * execution strategy. After this action finishes, any finalizers added
+     * to this scope will be run immediately.
+     */
+    def releaseAll(exit: Exit[Any, Any], execStrategy: ExecutionStrategy): UIO[Any]
 
     /**
-     * Runs the finalizers associated with this scope sequentially. After this
-     * any finalizers added to this scope will be run immediately.
+     * Removes the finalizer associated with this key and returns it.
      */
-    def runAll(exit: Exit[Any, Any]): URIO[R, Unit]
+    def remove(key: ReleaseKey): UIO[Option[Finalizer]]
 
     /**
-     * Runs the finalizers associated with this scope in parallel. After this
-     * any finalizers added to this scope will be run immediately.
+     * Replaces the finalizer associated with this key and returns it.
+     * If the finalizers associated with this scope have already been run this
+     * finalizer will be run immediately.
      */
-    def runAllPar(exit: Exit[Any, Any]): URIO[R, Unit]
+    def replace(key: ReleaseKey, finalizer: Finalizer): UIO[Option[Finalizer]]
   }
 
-  object FinalizerRef {
+  object ReleaseMap {
+    private[zio] def makeManaged(executionStrategy: ExecutionStrategy): ZManaged[Any, Nothing, ReleaseMap] =
+      ZManaged.makeExit(make)((m, e) => m.releaseAll(e, executionStrategy))
 
     /**
-     * Constructs a new `FinalizerRef` with the specified initial finalizer.
+     * Creates a new ReleaseMap.
      */
-    def make[R](initial: Exit[Any, Any] => URIO[R, Any]): UIO[FinalizerRef[R]] = {
+    def make: UIO[ReleaseMap] = {
+      sealed abstract class State
+      final case class Exited(nextKey: ReleaseKey, exit: Exit[Any, Any], done: Map[ReleaseKey, Finalizer]) extends State
+      final case class Running(nextKey: ReleaseKey, finalizers: Map[ReleaseKey, Finalizer])                extends State
+
+      Ref.make[State](Running(ReleaseKey.initial, Map())).map { ref =>
```

Comment:
What about using a TreeSeqMap or other insertion order preserving data structure here?

Reply:
That looks really nice. I wasn't aware of that.

But from what I see, it's only available for 2.13 :-(

Reply:
I faced a similar problem in keeper some time ago. I was thinking / planning to reimplement Java's LinkedHashMap (or some alternative) as a low-level STM data structure. Might also be a good fit here

Reply:
Well, if you'd be up for that in the future, it's going to be very useful as we're planning on baking `ReleaseMap` (in some form) into the `FiberContext`. So a more efficient implementation would be awesome!

Reply:
For this PR, I think the current scheme is sufficient, and we can improve on it later on 💪🏻

Reply:
@mschuwalow So I ended up switching to `scala.collection.immutable.LongMap`. Cut down a nice chunk of logic and also keeps iteration order for us 💪🏻

Reply:
very cool stuff!